presto-hive/src/main/java/com/facebook/presto/hive/HiveTableLayoutHandle.java

```diff
@@ -270,15 +274,30 @@ public Object getIdentifier(Optional<ConnectorSplit> split)
             domainPredicate = TupleDomain.fromColumnDomains(Optional.of(columnDomains));
         }
 
+        // Canonicalize constraint by removing constants when column is a partition key. This assumes
+        // all partitions are similar, and will have similar statistics like size, cardinality etc.
+        constraint = canonicalizationStrategy == PlanCanonicalizationStrategy.DEFAULT ?
+                TupleDomain.all() :
+                withColumnDomains(constraint.getDomains().get().entrySet().stream()
+                        .sorted(comparing(entry -> entry.getKey().toString()))
+                        .collect(toLinkedMap(Map.Entry::getKey, entry -> entry.getValue().canonicalize(isPartitionKey(entry.getKey())))));
```

Comment:
I stepped through your test `testCanonicalizationStrategies` and see if we can avoid passing `TupleDomain<ColumnHandle> constraint` around. Actually, I found `partitionColumnPredicate` is a perfect thing you can use directly from this class. Then you don't have to add `constraint` for `getIdentifier` interface.

Reply:
I tried this and one of my tests failed. These are different things. 

Comparing `ds > '<DATEID-1>'` and `ds > '<DATEID-10>'`

After removing constants, partitionColumnPredicate will make them both same, as it just contains the predicate.
Constraint on table however, will mark them differently as it will have a list of actual partitions

Reply:
Can we figure out what changed between table scan and partitionColumnPredicate? Also, it is fairly dangerous to have the assumption that table constraints will always be a list of actual values.... In this case, probably we don't want to replace the constants.... In general, I feel ranges are risky compared to single values.

Reply:
I checked the code. It is coming from `ConnectorTableLayout`, which further comes from `HiveMetadata.getTableLayout`. I feel this is too risky an approach to rely on value lists as it is internal implementation of HiveMetadata. Can we skip canonization for range predicate for now?

Reply:
This may be a problem for "ts" columns, but we can have different implementation in prism for that. I like this idea to skip removing constants for range predicates, and keep it if its a list. Going to code it here.

I wanted to add another strategy to remove constants in everything except range predicates in whole query. WIll implement it in future.

Reply:
Skipped Remove constants from range predicates, and cleaned up the interface